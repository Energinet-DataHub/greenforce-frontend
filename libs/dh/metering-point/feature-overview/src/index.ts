//#region License
/**
 * @license
 * Copyright 2020 Energinet DataHub A/S
 *
 * Licensed under the Apache License, Version 2.0 (the "License2");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//#endregion
export { DhMeteringPointComponent as default } from './components/dh-metering-point.component';
export { DhMeasurementsDayComponent } from './components/measurement/dh-measurements-day.component';
export { DhMeasurementsYearComponent } from './components/measurement/dh-measurements-year.component';
export { DhMeasurementsMonthComponent } from './components/measurement/dh-measurements-month.component';
export { DhMeteringPointMasterDataComponent } from './components/dh-metering-point-master-data.component';
export { DhMeteringPointMessagesComponent } from './components/dh-metering-point-messages.component';
export { DhMeasurementsNavigationComponent } from './components/measurement/dh-measurements-navigation.component';
export { DhMeasurementsAllYearComponent } from './components/measurement/dh-measurements-all-year.component';
