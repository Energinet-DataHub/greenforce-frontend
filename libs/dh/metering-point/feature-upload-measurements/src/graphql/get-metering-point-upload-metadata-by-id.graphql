query GetMeteringPointUploadMetadataById($meteringPointId: String!) {
  meteringPoint(meteringPointId: $meteringPointId) {
    id
    metadata {
      id
      measureUnit
      resolution
      type
      subType
    }
  }
}
