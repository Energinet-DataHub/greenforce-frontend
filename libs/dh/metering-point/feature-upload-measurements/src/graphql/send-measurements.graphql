mutation SendMeasurements($input: SendMeasurementsInput!) {
  sendMeasurements(input: $input)
}
