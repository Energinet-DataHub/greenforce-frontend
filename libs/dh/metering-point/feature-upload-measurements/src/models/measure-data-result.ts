//#region License
/**
 * @license
 * Copyright 2020 Energinet DataHub A/S
 *
 * Licensed under the Apache License, Version 2.0 (the "License2");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//#endregion
import { dayjs } from '@energinet/watt/date';
import {
  SendMeasurementsQuality,
  SendMeasurementsResolution,
} from '@energinet-datahub/dh/shared/domain/graphql';
import { danishTimeZoneIdentifier } from '@energinet/watt/datepicker';

export type Measurement = {
  position: number;
  quantity: number;
  quality: SendMeasurementsQuality;
};

export type MeasureDataParseError = {
  key: string;
  index: number;
};

const dateFormatInZone = new Intl.DateTimeFormat('da-DK', {
  timeZone: 'Europe/Copenhagen',
  dateStyle: 'short',
  timeStyle: 'short',
});

const dateFormat = new Intl.DateTimeFormat('da-DK', {
  timeZone: 'UTC',
  dateStyle: 'short',
  timeStyle: 'short',
});

export class MeasureDataResult {
  // List of supported date time formats
  private readonly DATETIME_FORMATS = [
    'D.M.YYYY H.mm', // 28.4.2025 0.00
    'DD.MM.YYYY H.mm', // 28.04.2025 0.00
    'YYYY-MM-DD H.mm', // 2025-04-28 0.15
    'YYYY-M-D H.mm', // 2025-4-28 0.15
    'D.M.YYYY HH.mm', // 28.4.2025 00.00
    'DD.MM.YYYY HH.mm', // 28.04.2025 00.00
    'YYYY-MM-DD HH.mm', // 2025-04-28 00.15
  ];

  /** Date of the first measurement. */
  first?: dayjs.Dayjs;

  /** Date of the last measurement. */
  last?: dayjs.Dayjs;

  /** Total sum of all measurements. */
  sum = 0;

  /** Progress of the parsing process. */
  progress = 0;

  /** Current row index. */
  index = 0;

  /** Whether the parsing process has encountered a fatal error. */
  isFatal = false;

  /** Set of qualities associated with the measurements. */
  qualities = new Set<SendMeasurementsQuality>();

  /** Array of measurements parsed from the CSV file. */
  measurements: Measurement[] = [];

  /** Array of errors encountered during parsing. */
  errors: MeasureDataParseError[] = [];

  /** Sets the progress to 100%. Chainable. */
  done = () => ((this.progress = 100), this);

  /** Adds an error to the list of errors. Chainable. */
  error = (key: string) => (this.errors.push({ key, index: this.index }), this);

  /** Marks the result as fatal and adds an error to the list of errors. Chainable. */
  fatal = (key: string) => ((this.isFatal = true), this.error(key));

  /** Try to set the current period, returns false if date is invalid. */
  trySetPeriod(period: string) {
    // Parse in local time, since `tz` is VERY slow.
    const date = dayjs(period, this.DATETIME_FORMATS);
    if (!date.isValid()) return false;

    // Correct the offset by keeping local time, adjusting the offset instead. Then save in UTC.
    const toZoned = () => date.tz(danishTimeZoneIdentifier, true).utc();
    if (!this.first) this.first = toZoned();

    // Adjust for DST
    // ----------------------------------------------------------------------------------------
    // When parsing a Date from the measurements CSV, specifically around DST boundaries,
    // the actual time can be ambiguous and can only be determined by looking at prior entries.
    // To parse this correctly requires a series of steps outlined below. Due to numerous bugs
    // found in Dayjs related to DST, the code uses `Intl` instead of utilities from Dayjs.
    // ----------------------------------------------------------------------------------------

    // Get the Date that `period` is supposed to parse to (if the data is correct). When parsing a
    // period within the DST transition, this Date may be different from the parsed Date, even when
    // the data is correct. BUT the dates will be equal when formatted (without offset), so this
    // can be used to adjust the period.
    const end = this.maybeGetEnd();
    if (!end) {
      this.last = toZoned();
    } else {
      // The end Date (which is in UTC) is formatted to a danish format, adjusting the hours
      // appropriately (fx. "2025-10-26T00:00:00.000Z" becomes "26.10.2025, 02.00").
      const expected = dateFormatInZone.format(end.toDate());

      // Remove the offset from the parsed Date (which can be in any local time), converting it
      // to UTC (fx. "2025-10-26T02:00:00.000+02:00" becomes "2025-10-26T02:00:00.000Z"). The Date
      // is now certainly wrong, but that is intentional. It can now be formatted similar to above,
      // but without timezone adjustment. The same example should then become "26.10.2025, 02.00".
      const actual = dateFormat.format(date.utc(true).toDate());

      // Finally, if the textual representations of the expected Date and the actual parsed Date
      // are equal, then correct the parsed date to be the same as the expected Date (end).
      this.last = expected === actual ? end : toZoned();
    }

    return true;
  }

  /** Returns the end date of all measurements or `null` if there no measurements. */
  maybeGetEnd = () => {
    if (!this.last) return null;
    switch (this.resolution) {
      case SendMeasurementsResolution.QuarterHourly:
        return this.last.add(15, 'minutes');
      case SendMeasurementsResolution.Hourly:
        return this.last.add(1, 'hour');
      case SendMeasurementsResolution.Monthly:
        // HACK: dayjs does not add days correctly across DST boundaries
        // https://github.com/iamkun/dayjs/issues/1271
        return this.last
          .tz(danishTimeZoneIdentifier)
          .add(1, 'month')
          .tz(danishTimeZoneIdentifier, true)
          .utc();
    }
  };

  /** Gets the current start and end of measurements. */
  maybeGetDateRange = () => {
    const start = this.first?.toDate();
    const end = this.maybeGetEnd()?.toDate();
    return start ? { start, end } : null;
  };

  /** Updates the progress of the parser based on the current cursor position. */
  updateProgress = (cursor: number) => {
    this.progress = Math.round(Math.min(cursor / this.fileSize, 1) * 100);
  };

  /** Adds a measurement to the result. Chainable. */
  addMeasurement = (measurement: Measurement) => {
    this.measurements.push(measurement);
    this.sum += measurement.quantity;
    this.qualities.add(measurement.quality);
    return this;
  };

  constructor(
    private fileSize: number,
    public resolution: SendMeasurementsResolution
  ) {}
}
