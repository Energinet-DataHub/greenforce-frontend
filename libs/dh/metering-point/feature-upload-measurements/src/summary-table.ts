//#region License
/**
 * @license
 * Copyright 2020 Energinet DataHub A/S
 *
 * Licensed under the Apache License, Version 2.0 (the "License2");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//#endregion
import { Component, input } from '@angular/core';
import { TranslocoDirective } from '@jsverse/transloco';

@Component({
  selector: 'dh-upload-measurements-summary-table',
  imports: [TranslocoDirective],
  styles: `
    @use '@energinet/watt/utils' as watt;

    table {
      width: 100%;
      border-collapse: collapse;

      & tr {
        height: 44px;
      }

      & th {
        text-align: left;
        padding-left: var(--watt-space-m);
      }

      & td {
        text-align: right;
        padding-right: var(--watt-space-m);
      }

      & th {
        @include watt.typography-watt-label;
      }

      & tfoot tr {
        background: var(--watt-color-primary-ultralight);
      }

      & tfoot td {
        @include watt.typography-font-weight('semi-bold');
      }
    }
  `,
  template: `
    <table *transloco="let t; prefix: 'meteringPoint.measurements'">
      <tbody>
        <tr>
          <th>{{ t('upload.table.positionCount') }}</th>
          <td>{{ positions() }}</td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <th>{{ t('upload.table.sum') }}</th>
          <td>{{ sum() }}</td>
        </tr>
      </tfoot>
    </table>
  `,
})
export class DhUploadMeasurementsSummaryTable {
  readonly positions = input.required<number>();
  readonly sum = input.required<number>();
}
