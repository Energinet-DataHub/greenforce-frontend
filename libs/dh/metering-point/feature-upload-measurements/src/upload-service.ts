//#region License
/**
 * @license
 * Copyright 2020 Energinet DataHub A/S
 *
 * Licensed under the Apache License, Version 2.0 (the "License2");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//#endregion
import { effect, Injectable } from '@angular/core';
import {
  ElectricityMarketMeteringPointType,
  MeteringPointType2,
  SendMeasurementsDocument,
  SendMeasurementsResolution,
} from '@energinet-datahub/dh/shared/domain/graphql';
import { MeasureDataResult } from './models/measure-data-result';
import { injectToast } from '@energinet-datahub/dh/shared/ui-util';
import { mutation } from '@energinet-datahub/dh/shared/util-apollo';
import { assert, assertIsDefined } from '@energinet-datahub/dh/shared/util-assert';
import { parseMeasurements } from './parse-measurements';

@Injectable({ providedIn: 'root' })
export class DhUploadMeasurementsService {
  private toast = injectToast('meteringPoint.measurements.upload.toast');
  private sendMeasurements = mutation(SendMeasurementsDocument);
  protected toastEffect = effect(() => this.toast(this.sendMeasurements.status()));

  /** Maps resolution string to SendMeasurementsResolution. */
  private mapResolution = (resolution: string) => {
    switch (resolution) {
      case 'PT15M':
        return SendMeasurementsResolution.QuarterHourly;
      case 'PT1H':
        return SendMeasurementsResolution.Hourly;
      case 'P1M':
        return SendMeasurementsResolution.Monthly;
      default:
        throw new Error(`Unsupported resolution: ${resolution}`);
    }
  };

  /** Maps metering point type to more narrow type.  */
  private mapMeteringPointType = (type: ElectricityMarketMeteringPointType) => {
    switch (type) {
      case ElectricityMarketMeteringPointType.Consumption:
        return MeteringPointType2.Consumption;
      case ElectricityMarketMeteringPointType.Production:
        return MeteringPointType2.Production;
      case ElectricityMarketMeteringPointType.Exchange:
        return MeteringPointType2.Exchange;
      case ElectricityMarketMeteringPointType.VeProduction:
        return MeteringPointType2.VeProduction;
      case ElectricityMarketMeteringPointType.Analysis:
        return MeteringPointType2.Analysis;
      default:
        throw new Error(`Unsupported metering point type: ${type}`);
    }
  };

  /** Parses a CSV file of measurement data, streaming the result. */
  parseFile = (file: File, resolution: string) =>
    parseMeasurements(file, this.mapResolution(resolution));

  /** Sends measurements to the server. */
  send = (
    meteringPointId: string,
    meteringPointType: ElectricityMarketMeteringPointType,
    result: MeasureDataResult
  ) => {
    const interval = result.maybeGetDateRange();
    assertIsDefined(interval?.start);
    assertIsDefined(interval?.end);
    assert(!result.errors.length);
    this.sendMeasurements.mutate({
      variables: {
        input: {
          meteringPointId,
          meteringPointType: this.mapMeteringPointType(meteringPointType),
          resolution: result.resolution,
          start: interval.start,
          end: interval.end,
          measurements: result.measurements,
        },
      },
    });
  };
}
