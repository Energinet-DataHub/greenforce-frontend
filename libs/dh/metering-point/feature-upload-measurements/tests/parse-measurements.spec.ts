/* eslint-disable sonarjs/no-duplicate-string */
//#region License
/**
 * @license
 * Copyright 2020 Energinet DataHub A/S
 *
 * Licensed under the Apache License, Version 2.0 (the "License2");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//#endregion
import { SendMeasurementsResolution } from '@energinet-datahub/dh/shared/domain/graphql';
import { parseMeasurements } from '../src/parse-measurements';
import { lastValueFrom } from 'rxjs';
import { MeasureDataResult } from '../src/models/measure-data-result';

const makeReadable = (result: MeasureDataResult) => ({
  interval: result.maybeGetDateRange(),
  errors: result.errors,
  isFatal: result.isFatal,
  measurements: result.measurements.length,
  progress: result.progress,
  qualities: [...result.qualities].join('/'),
  resolution: result.resolution,
  sum: result.sum,
});

describe(parseMeasurements, () => {
  it('should parse measurements in quarter hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,28.4.2025 0.00,2,Målt',
      '2,28.4.2025 0.15,2,Målt',
      '3,28.4.2025 0.30,2,Målt',
      '4,28.4.2025 0.45,2,Målt',
      '5,28.4.2025 1.00,2,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.QuarterHourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should parse measurements in hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,28.4.2025 0.00,8,Målt',
      '2,28.4.2025 1.00,8,Målt',
      '3,28.4.2025 2.00,8,Målt',
      '4,28.4.2025 3.00,8,Målt',
      '5,28.4.2025 4.00,8,Estimeret',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Hourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should parse measurements in monthly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,1.1.2025 0.00,100,Estimeret',
      '2,1.2.2025 0.00,100,Målt',
      '3,1.3.2025 0.00,100,Målt',
      '4,1.4.2025 0.00,100,Målt',
      '5,1.5.2025 0.00,100,Målt',
      '6,1.6.2025 0.00,100,Målt',
      '7,1.7.2025 0.00,100,Målt',
      '8,1.8.2025 0.00,100,Målt',
      '9,1.9.2025 0.00,100,Målt',
      '10,1.10.2025 0.00,100,Målt',
      '11,1.11.2025 0.00,100,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Monthly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should handle DST start correctly in quarter hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,30.3.2025 0.00,2,Målt',
      '2,30.3.2025 0.15,2,Målt',
      '3,30.3.2025 0.30,2,Målt',
      '4,30.3.2025 0.45,2,Målt',
      '5,30.3.2025 1.00,2,Målt',
      '6,30.3.2025 1.15,2,Estimeret',
      '7,30.3.2025 1.30,2,Målt',
      '8,30.3.2025 1.45,2,Målt',
      '9,30.3.2025 3.00,2,Målt',
      '10,30.3.2025 3.15,2,Målt',
      '11,30.3.2025 3.30,2,Målt',
      '12,30.3.2025 3.45,2,Målt',
      '13,30.3.2025 4.00,2,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.QuarterHourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should handle DST start correctly in hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,30.3.2025 0.00,2,Målt',
      '2,30.3.2025 1.00,2,Målt',
      '3,30.3.2025 3.00,2,Målt',
      '4,30.3.2025 4.00,2,Målt',
      '5,30.3.2025 5.00,2,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Hourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should handle DST end correctly in quarter hourly resolution', async () => {
    // we might have a problem if running the parser on computers that are not using
    // the danish timezone, due to DST...
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,26.10.2025 0.00,2,Målt',
      '2,26.10.2025 0.15,2,Målt',
      '3,26.10.2025 0.30,2,Målt',
      '4,26.10.2025 0.45,2,Målt',
      '5,26.10.2025 1.00,2,Målt',
      '6,26.10.2025 1.15,2,Estimeret',
      '7,26.10.2025 1.30,2,Målt',
      '8,26.10.2025 1.45,2,Målt',
      '9,26.10.2025 2.00,2,Målt',
      '10,26.10.2025 2.15,2,Målt',
      '11,26.10.2025 2.30,2,Målt',
      '12,26.10.2025 2.45,2,Målt',
      '13,26.10.2025 2.00,2,Målt',
      '14,26.10.2025 2.15,2,Målt',
      '15,26.10.2025 2.30,2,Målt',
      '16,26.10.2025 2.45,2,Målt',
      '17,26.10.2025 3.00,2,Målt',
      '18,26.10.2025 3.15,2,Målt',
      '19,26.10.2025 3.30,2,Målt',
      '20,26.10.2025 3.45,2,Målt',
      '21,26.10.2025 4.00,2,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.QuarterHourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should handle DST end correctly in hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,26.10.2025 0.00,2,Målt',
      '2,26.10.2025 1.00,2,Målt',
      '3,26.10.2025 2.00,2,Målt',
      '4,26.10.2025 2.00,2,Målt',
      '5,26.10.2025 3.00,2,Målt',
      '6,26.10.2025 4.00,2,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Hourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should error with missing measurements in quarter hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,28.4.2025 0.00,2,Målt',
      '3,28.4.2025 0.30,2,Målt',
      '4,28.4.2025 0.45,2,Målt',
      '5,28.4.2025 1.00,2,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.QuarterHourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should error with unexpected measurements in quarter hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,28.4.2025 0.00,2,Målt',
      '2,28.4.2025 0.05,2,Målt',
      '3,28.4.2025 0.15,2,Målt',
      '4,28.4.2025 0.30,2,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.QuarterHourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should error with missing measurements in hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,28.4.2025 0.00,8,Målt',
      '2,28.4.2025 2.00,8,Målt',
      '3,28.4.2025 3.00,8,Målt',
      '4,28.4.2025 4.00,8,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Hourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should error with unexpected measurements in hourly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,28.4.2025 0.00,8,Målt',
      '2,28.4.2025 0.15,8,Målt',
      '3,28.4.2025 2.00,8,Målt',
      '4,28.4.2025 3.00,8,Målt',
      '5,28.4.2025 4.00,8,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Hourly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should error with missing measurements in monthly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,1.1.2025 0.00,100,Estimeret',
      '2,1.2.2025 0.00,100,Målt',
      '3,1.4.2025 0.00,100,Målt',
      '4,1.5.2025 0.00,100,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Monthly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });

  it('should error with unexpected measurements in monthly resolution', async () => {
    const csv = [
      'Position,Periode,Værdi,Kvantum status',
      '1,1.1.2025 0.00,100,Estimeret',
      '2,1.2.2025 0.00,100,Målt',
      '3,1.3.2025 0.00,100,Målt',
      '4,1.4.2025 0.00,100,Målt',
      '5,1.5.2025 0.00,100,Målt',
      '6,1.5.2025 0.00,100,Målt',
    ].join('\n');

    const stream = parseMeasurements(csv, SendMeasurementsResolution.Monthly);
    const result = await lastValueFrom(stream);
    expect(makeReadable(result)).toMatchSnapshot();
  });
});

// case 'PT15M':
//   return SendMeasurementsResolution.QuarterHourly;
// case 'PT1H':
//   return SendMeasurementsResolution.Hourly;
// case 'P1M':
