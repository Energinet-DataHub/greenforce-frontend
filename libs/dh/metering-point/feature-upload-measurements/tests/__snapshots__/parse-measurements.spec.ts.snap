// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseMeasurements should error when period is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 0,
      "key": "INVALID_PERIOD",
    },
  ],
  "interval": null,
  "isFatal": true,
  "measurements": 0,
  "progress": 42,
  "qualities": "",
  "resolution": "QUARTER_HOURLY",
  "sum": 0,
}
`;

exports[`parseMeasurements should error when position is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 0,
      "key": "INVALID_POSITION",
    },
  ],
  "interval": Object {
    "end": 2025-10-25T23:15:00.000Z,
    "start": 2025-10-25T22:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 4,
  "progress": 100,
  "qualities": "MEASURED",
  "resolution": "QUARTER_HOURLY",
  "sum": 8,
}
`;

exports[`parseMeasurements should error when quality is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 3,
      "key": "INVALID_QUALITY",
    },
  ],
  "interval": Object {
    "end": 2025-10-25T23:15:00.000Z,
    "start": 2025-10-25T22:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 4,
  "progress": 100,
  "qualities": "MEASURED/ESTIMATED",
  "resolution": "QUARTER_HOURLY",
  "sum": 8,
}
`;

exports[`parseMeasurements should error when quantity is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 0,
      "key": "INVALID_QUANTITY",
    },
  ],
  "interval": Object {
    "end": 2025-10-25T23:15:00.000Z,
    "start": 2025-10-25T22:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 4,
  "progress": 100,
  "qualities": "MEASURED",
  "resolution": "QUARTER_HOURLY",
  "sum": 8,
}
`;

exports[`parseMeasurements should error when structure is unexpected 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 0,
      "key": "STRUCTURE",
    },
  ],
  "interval": null,
  "isFatal": true,
  "measurements": 0,
  "progress": 33,
  "qualities": "",
  "resolution": "MONTHLY",
  "sum": 0,
}
`;

exports[`parseMeasurements should error with missing measurements in hourly resolution 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 1,
      "key": "MISSING_MEASUREMENT",
    },
  ],
  "interval": Object {
    "end": 2025-04-28T01:00:00.000Z,
    "start": 2025-04-27T22:00:00.000Z,
  },
  "isFatal": true,
  "measurements": 1,
  "progress": 65,
  "qualities": "MEASURED",
  "resolution": "HOURLY",
  "sum": 8,
}
`;

exports[`parseMeasurements should error with missing measurements in monthly resolution 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 2,
      "key": "MISSING_MEASUREMENT",
    },
  ],
  "interval": Object {
    "end": 2025-04-30T22:00:00.000Z,
    "start": 2024-12-31T23:00:00.000Z,
  },
  "isFatal": true,
  "measurements": 2,
  "progress": 83,
  "qualities": "ESTIMATED/MEASURED",
  "resolution": "MONTHLY",
  "sum": 200,
}
`;

exports[`parseMeasurements should error with missing measurements in quarter hourly resolution 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 1,
      "key": "MISSING_MEASUREMENT",
    },
  ],
  "interval": Object {
    "end": 2025-04-27T22:45:00.000Z,
    "start": 2025-04-27T22:00:00.000Z,
  },
  "isFatal": true,
  "measurements": 1,
  "progress": 65,
  "qualities": "MEASURED",
  "resolution": "QUARTER_HOURLY",
  "sum": 2,
}
`;

exports[`parseMeasurements should error with unexpected measurements in hourly resolution 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 1,
      "key": "UNEXPECTED_MEASUREMENT",
    },
  ],
  "interval": Object {
    "end": 2025-04-27T23:15:00.000Z,
    "start": 2025-04-27T22:00:00.000Z,
  },
  "isFatal": true,
  "measurements": 1,
  "progress": 55,
  "qualities": "MEASURED",
  "resolution": "HOURLY",
  "sum": 8,
}
`;

exports[`parseMeasurements should error with unexpected measurements in monthly resolution 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 5,
      "key": "UNEXPECTED_MEASUREMENT",
    },
  ],
  "interval": Object {
    "end": 2025-05-31T22:00:00.000Z,
    "start": 2024-12-31T23:00:00.000Z,
  },
  "isFatal": true,
  "measurements": 5,
  "progress": 100,
  "qualities": "ESTIMATED/MEASURED",
  "resolution": "MONTHLY",
  "sum": 500,
}
`;

exports[`parseMeasurements should error with unexpected measurements in quarter hourly resolution 1`] = `
Object {
  "errors": Array [
    Object {
      "index": 1,
      "key": "UNEXPECTED_MEASUREMENT",
    },
  ],
  "interval": Object {
    "end": 2025-04-27T22:20:00.000Z,
    "start": 2025-04-27T22:00:00.000Z,
  },
  "isFatal": true,
  "measurements": 1,
  "progress": 65,
  "qualities": "MEASURED",
  "resolution": "QUARTER_HOURLY",
  "sum": 2,
}
`;

exports[`parseMeasurements should handle DST end correctly in hourly resolution 1`] = `
Object {
  "errors": Array [],
  "interval": Object {
    "end": 2025-10-26T04:00:00.000Z,
    "start": 2025-10-25T22:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 6,
  "progress": 100,
  "qualities": "MEASURED",
  "resolution": "HOURLY",
  "sum": 12,
}
`;

exports[`parseMeasurements should handle DST end correctly in quarter hourly resolution 1`] = `
Object {
  "errors": Array [],
  "interval": Object {
    "end": 2025-10-26T03:15:00.000Z,
    "start": 2025-10-25T22:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 21,
  "progress": 100,
  "qualities": "MEASURED/ESTIMATED",
  "resolution": "QUARTER_HOURLY",
  "sum": 42,
}
`;

exports[`parseMeasurements should handle DST start correctly in hourly resolution 1`] = `
Object {
  "errors": Array [],
  "interval": Object {
    "end": 2025-03-30T04:00:00.000Z,
    "start": 2025-03-29T23:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 5,
  "progress": 100,
  "qualities": "MEASURED",
  "resolution": "HOURLY",
  "sum": 10,
}
`;

exports[`parseMeasurements should handle DST start correctly in quarter hourly resolution 1`] = `
Object {
  "errors": Array [],
  "interval": Object {
    "end": 2025-03-30T02:15:00.000Z,
    "start": 2025-03-29T23:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 13,
  "progress": 100,
  "qualities": "MEASURED/ESTIMATED",
  "resolution": "QUARTER_HOURLY",
  "sum": 26,
}
`;

exports[`parseMeasurements should parse measurements in hourly resolution 1`] = `
Object {
  "errors": Array [],
  "interval": Object {
    "end": 2025-04-28T03:00:00.000Z,
    "start": 2025-04-27T22:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 5,
  "progress": 100,
  "qualities": "MEASURED/ESTIMATED",
  "resolution": "HOURLY",
  "sum": 40,
}
`;

exports[`parseMeasurements should parse measurements in monthly resolution 1`] = `
Object {
  "errors": Array [],
  "interval": Object {
    "end": 2025-11-30T23:00:00.000Z,
    "start": 2024-12-31T23:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 11,
  "progress": 100,
  "qualities": "ESTIMATED/MEASURED",
  "resolution": "MONTHLY",
  "sum": 1100,
}
`;

exports[`parseMeasurements should parse measurements in quarter hourly resolution 1`] = `
Object {
  "errors": Array [],
  "interval": Object {
    "end": 2025-04-27T23:15:00.000Z,
    "start": 2025-04-27T22:00:00.000Z,
  },
  "isFatal": false,
  "measurements": 5,
  "progress": 100,
  "qualities": "MEASURED",
  "resolution": "QUARTER_HOURLY",
  "sum": 10,
}
`;
